# Data Preprocessing

This notebook focuses on preparing the customer churn dataset for machine learning.

Objectives:

- Remove irrelevant features
- Handle missing values
- Handle duplicate records
- Convert categorical variables into numerical form
- Prepare data for model training
- Prevent data leakage
- Create a clean machine learning dataset

In [25]:
# Import required libraries

import pandas as pd
import numpy as np

# Load dataset

df = pd.read_csv(
    "../data/WA_Fn-UseC_-Telco-Customer-Churn.csv"
)

# Display shape

print("Dataset Shape:", df.shape)

# Display first rows

df.head()

Dataset Shape: (7043, 21)


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


# Dataset Information

Inspect dataset structure, data types, and missing values before preprocessing.

In [26]:
# Display dataset information

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


# Check Duplicate Records

Duplicate records can negatively affect model performance.

This step identifies duplicate observations within the dataset.

In [27]:
# Count duplicate rows

duplicate_count = df.duplicated().sum()

# Display duplicate count

print("Duplicate Records:", duplicate_count)

Duplicate Records: 0


# Check Unique Customer IDs

Customer ID should uniquely identify each customer.

This step verifies uniqueness before removing the identifier column.

In [28]:
# Count unique customer IDs

unique_customers = df["customerID"].nunique()

# Total records

total_records = len(df)

# Display results

print("Unique Customer IDs:", unique_customers)

print("Total Records:", total_records)

Unique Customer IDs: 7043
Total Records: 7043


# Remove Customer Identifier

Customer ID is a unique identifier and does not contain predictive information.

Keeping customerID may introduce noise and reduce model generalization.

Therefore, the column will be removed before model training.

In [29]:
# Remove customerID column

df = df.drop(
    columns=["customerID"]
)

# Verify removal

df.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


# Observation

- No duplicate records were found in the dataset.
- customerID contains unique values for every customer.
- customerID does not contribute useful information for churn prediction.
- The column was removed to improve model quality and prevent unnecessary complexity.

# TotalCharges Data Type Investigation

The TotalCharges column is currently stored as an object data type.

This is unexpected because TotalCharges represents monetary values and should be numerical.

The objective of this step is to identify invalid values and convert the column into a numeric format.

In [30]:
# Check current data type

print(df["TotalCharges"].dtype)

# Display first values

df["TotalCharges"].head()

object


0      29.85
1     1889.5
2     108.15
3    1840.75
4     151.65
Name: TotalCharges, dtype: object

# Identify Invalid TotalCharges Values

Blank spaces inside the TotalCharges column prevent proper numerical conversion.

This step counts how many invalid records exist.

In [31]:
# Count blank values

blank_count = (df["TotalCharges"] == " ").sum()

# Display result

print("Blank Values in TotalCharges:", blank_count)

Blank Values in TotalCharges: 11


# Convert TotalCharges to Numeric

Invalid values will be converted into missing values (NaN).

This allows numerical processing and statistical analysis.

In [32]:
# Convert TotalCharges to numeric

df["TotalCharges"] = pd.to_numeric(
    df["TotalCharges"],
    errors="coerce"
)

# Verify data type

print(df["TotalCharges"].dtype)

float64


# Missing Value Audit

After conversion, blank entries become missing values (NaN).

This step identifies the total number of missing values present in each column.

In [33]:
# Check missing values

df.isnull().sum()

gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

# Handle Missing Values

Since TotalCharges contains only a small number of missing values, median imputation will be used.

Median is robust against skewed distributions and outliers.

In [34]:
# Fill missing values using median

df["TotalCharges"] = df["TotalCharges"].fillna(
    df["TotalCharges"].median()
)

# Verify missing values

df.isnull().sum()

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

# Observation

- TotalCharges was incorrectly stored as an object data type.
- Blank values were identified and converted to missing values.
- The column was successfully converted to a numerical format.
- Missing values were handled using median imputation.
- The dataset is now ready for feature encoding and machine learning preparation.

# Preprocessing Audit Report

This section summarizes all preprocessing actions performed on the dataset.

The objective is to maintain transparency, reproducibility, and data quality throughout the machine learning pipeline.

# Dataset Quality Summary

## Original Dataset

- Records: 7043
- Features: 21

## Duplicate Check

- Duplicate Records Found: 0
- Action Taken: No records removed

## Customer Identifier

- customerID identified as a unique identifier
- Action Taken: Removed from dataset

## TotalCharges Column

- Originally stored as object datatype
- 11 blank values detected
- Converted to numeric datatype
- Missing values handled using median imputation

## Final Dataset Status

- No duplicate records
- No missing values
- Correct numerical data types

- Ready for feature encoding

# Feature Categorization

Before encoding, features are categorized based on the number of unique values.

This helps apply the most appropriate encoding technique to each feature.

In [35]:
# Display unique values for each column

for column in df.columns:
    
    print("\n")
    
    print(column)
    
    print(df[column].unique())



gender
['Female' 'Male']


SeniorCitizen
[0 1]


Partner
['Yes' 'No']


Dependents
['No' 'Yes']


tenure
[ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 40 59 24 44 19 54 51 26  0
 39]


PhoneService
['No' 'Yes']


MultipleLines
['No phone service' 'No' 'Yes']


InternetService
['DSL' 'Fiber optic' 'No']


OnlineSecurity
['No' 'Yes' 'No internet service']


OnlineBackup
['Yes' 'No' 'No internet service']


DeviceProtection
['No' 'Yes' 'No internet service']


TechSupport
['No' 'Yes' 'No internet service']


StreamingTV
['No' 'Yes' 'No internet service']


StreamingMovies
['No' 'Yes' 'No internet service']


Contract
['Month-to-month' 'One year' 'Two year']


PaperlessBilling
['Yes' 'No']


PaymentMethod
['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']


MonthlyCharges
[29.85 56.95 53.85 ... 63.1  44.2


Many beginners wrongly merge them.

We won't.

---

# Markdown Cell

```markdown
# Encoding Strategy

The dataset contains both binary and multi-class categorical features.

Encoding approach:

## Binary Encoding

The following features will be converted into 0 and 1:

- gender
- Partner
- Dependents
- PhoneService
- PaperlessBilling
- Churn

## One-Hot Encoding

The following features contain multiple categories and will be transformed using One-Hot Encoding:

- MultipleLines
- InternetService
- OnlineSecurity
- OnlineBackup
- DeviceProtection
- TechSupport
- StreamingTV
- StreamingMovies
- Contract
- PaymentMethod

Special categories such as "No internet service" and "No phone service" will be preserved because they contain meaningful business information.

# Binary Feature Encoding

Binary categorical variables are converted into numerical format using label encoding.

This step prepares the data for machine learning algorithms.

In [36]:
# Encode binary categorical features

binary_mapping = {
    "Yes": 1,
    "No": 0
}

# Encode gender

df["gender"] = df["gender"].map({
    "Female": 0,
    "Male": 1
})

# Encode Partner

df["Partner"] = df["Partner"].map(binary_mapping)

# Encode Dependents

df["Dependents"] = df["Dependents"].map(binary_mapping)

# Encode PhoneService

df["PhoneService"] = df["PhoneService"].map(binary_mapping)

# Encode PaperlessBilling

df["PaperlessBilling"] = df["PaperlessBilling"].map(binary_mapping)

# Encode target variable

df["Churn"] = df["Churn"].map(binary_mapping)

# Verify results

df.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,0,0,1,0,1,0,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,1,Electronic check,29.85,29.85,0
1,1,0,0,0,34,1,No,DSL,Yes,No,Yes,No,No,No,One year,0,Mailed check,56.95,1889.50,0
2,1,0,0,0,2,1,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,1,Mailed check,53.85,108.15,1
3,1,0,0,0,45,0,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,0,Bank transfer (automatic),42.30,1840.75,0
4,0,0,0,0,2,1,No,Fiber optic,No,No,No,No,No,No,Month-to-month,1,Electronic check,70.70,151.65,1


# Observation

Binary categorical features were successfully converted into numerical format.

Encoding Summary:

- Female → 0
- Male → 1
- No → 0
- Yes → 1

The target variable (Churn) has also been converted into machine learning compatible format.

In [37]:
# Verify encoded columns

df[
    [
        "gender",
        "Partner",
        "Dependents",
        "PhoneService",
        "PaperlessBilling",
        "Churn"
    ]
].head()

,gender,Partner,Dependents,PhoneService,PaperlessBilling,Churn
0,0,1,0,0,1,0
1,1,0,0,1,0,0
2,1,0,0,1,1,1
3,1,0,0,0,0,0
4,0,0,0,1,1,1


# One-Hot Encoding

Several categorical features contain more than two categories.

These variables will be transformed using One-Hot Encoding to make them suitable for machine learning algorithms.

This approach prevents the model from assuming an incorrect ordinal relationship between categories.

In [38]:
# List multi-class categorical features

multiclass_features = [
    "MultipleLines",
    "InternetService",
    "OnlineSecurity",
    "OnlineBackup",
    "DeviceProtection",
    "TechSupport",
    "StreamingTV",
    "StreamingMovies",
    "Contract",
    "PaymentMethod"
]

# Apply One-Hot Encoding

df = pd.get_dummies(
    df,
    columns=multiclass_features,
    drop_first=True
)

# Display shape after encoding

print("Dataset Shape:", df.shape)

# Display first rows

df.head()

Dataset Shape: (7043, 31)


,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,PaperlessBilling,MonthlyCharges,TotalCharges,Churn,...,TechSupport_Yes,StreamingTV_No internet service,StreamingTV_Yes,StreamingMovies_No internet service,StreamingMovies_Yes,Contract_One year,Contract_Two year,PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check
0,0,0,1,0,1,0,1,29.85,29.85,0,...,False,False,False,False,False,False,False,False,True,False
1,1,0,0,0,34,1,0,56.95,1889.50,0,...,False,False,False,False,False,True,False,False,False,True
2,1,0,0,0,2,1,1,53.85,108.15,1,...,False,False,False,False,False,False,False,False,False,True
3,1,0,0,0,45,0,0,42.30,1840.75,0,...,True,False,False,False,False,True,False,False,False,False
4,0,0,0,0,2,1,1,70.70,151.65,1,...,False,False,False,False,False,False,False,False,True,False


# Observation

Multi-class categorical variables were transformed into numerical features using One-Hot Encoding.

Benefits:

- Prevents machine learning models from interpreting categories as ordered values.
- Retains important business information.
- Produces a fully numerical dataset suitable for model training.

The dataset is now significantly expanded because each category is represented by its own feature column.

# Feature Engineering

Feature Engineering creates new variables from existing data to improve predictive performance.

These engineered features capture customer value, loyalty, and subscription maturity, which are important indicators of churn behavior.

In [39]:
# Create average monthly spend

df["AverageMonthlySpend"] = (
    df["TotalCharges"] /
    (df["tenure"] + 1)
)

# Create high value customer flag

df["HighValueCustomer"] = (
    df["MonthlyCharges"] >
    df["MonthlyCharges"].median()
).astype(int)

# Create long-term customer flag

df["LongTermCustomer"] = (
    df["tenure"] >= 24
).astype(int)

# Create tenure groups

df["TenureGroup"] = pd.cut(
    df["tenure"],
    bins=[-1, 12, 24, 48, 72],
    labels=[
        "New",
        "Growing",
        "Established",
        "Loyal"
    ]
)

# Display sample

df[
    [
        "tenure",
        "MonthlyCharges",
        "TotalCharges",
        "AverageMonthlySpend",
        "HighValueCustomer",
        "LongTermCustomer",
        "TenureGroup"
    ]
].head()

,tenure,MonthlyCharges,TotalCharges,AverageMonthlySpend,HighValueCustomer,LongTermCustomer,TenureGroup
0,1,29.85,29.85,14.925000,0,0,New
1,34,56.95,1889.50,53.985714,0,1,Established
2,2,53.85,108.15,36.050000,0,0,New
3,45,42.30,1840.75,40.016304,0,1,Established
4,2,70.70,151.65,50.550000,1,0,New


# Observation

New business-oriented features were created:

- AverageMonthlySpend estimates spending behavior across customer lifetime.
- HighValueCustomer identifies customers paying above the median monthly charge.
- LongTermCustomer identifies customers retained for at least 24 months.
- TenureGroup categorizes customers based on subscription age.

These engineered features can improve model performance and provide stronger business insights.

# Encode Tenure Group

In [40]:
# One-hot encode tenure group

df = pd.get_dummies(
    df,
    columns=["TenureGroup"],
    drop_first=True,
    dtype=int
)

# Check shape

print(df.shape)

# Display first rows

df.head()

(7043, 37)


,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,PaperlessBilling,MonthlyCharges,TotalCharges,Churn,...,Contract_Two year,PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check,AverageMonthlySpend,HighValueCustomer,LongTermCustomer,TenureGroup_Growing,TenureGroup_Established,TenureGroup_Loyal
0,0,0,1,0,1,0,1,29.85,29.85,0,...,False,False,True,False,14.925000,0,0,0,0,0
1,1,0,0,0,34,1,0,56.95,1889.50,0,...,False,False,False,True,53.985714,0,1,0,1,0
2,1,0,0,0,2,1,1,53.85,108.15,1,...,False,False,False,True,36.050000,0,0,0,0,0
3,1,0,0,0,45,0,0,42.30,1840.75,0,...,False,False,False,False,40.016304,0,1,0,1,0
4,0,0,0,0,2,1,1,70.70,151.65,1,...,False,False,True,False,50.550000,1,0,0,0,0


# Observation

TenureGroup was transformed into numerical features using One-Hot Encoding.

The dataset now contains both original and engineered features suitable for machine learning model development.

# Save Processed Dataset

In [41]:
# Save processed dataset

df.to_csv(
    "../data/processed_telco_churn.csv",
    index=False
)

print("Processed dataset saved successfully.")

Processed dataset saved successfully.
